Add switches (on/off zones) to geniushub

castaway opened the pull request:
## Description:

The **geniushub** integration shows smart plug Devices as BinarySensors; it can report when they are on or off, but can not change their state (this is done in GH via the parent Zone). 

This PR adds on/off Zones as SwitchDevices:
 - 'turn_on' - set Zone to GH's Override/On mode
 - 'turn_off' - set Zone to GH's Timer mode

Also bumps **geniushub-client** from 0.6.28 to 0.6.29, see: https://github.com/zxdavb/geniushub-client/compare/v0.6.28...v0.630, which fixes:
 - https://github.com/zxdavb/geniushub-client/issues/25, and
 - https://github.com/zxdavb/geniushub-client/issues/27

**Pull request with documentation for home-assistant.io:** home-assistant/home-assistant.io#11095

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass** - no tests for geniushub
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`. - nothing changed
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`. - no new ones
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

castaway committed:
New switch for geniushub outlets

castaway committed:
Load new geniushub switch component

homeassistant commented:
Hi @castaway, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#28182).

Once you do that we will be able to review and accept this pull request.

Thanks!

castaway committed:
Remove useless "__init__" function

zxdavb commented on homeassistant/components/geniushub/switch.py:
```diff
@@ -0,0 +1,51 @@
+
+from . import DOMAIN, GeniusZone
+
+GH_ZONE_ATTR = "on / off"
```
`GH_ON_OFF_ZONE` please, see below.

zxdavb commented on homeassistant/components/geniushub/switch.py:
```diff
@@ -0,0 +1,51 @@
+        await self._zone.set_mode("off")
+        # State is set optimistically in the command above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state()
```
My understanding is that this is not required.

castaway replied:
I'm not sure what you mean by this (and the following comment re set_mode), not being required - these are the two methods by which we switch the zone on/off.

As far as I can tell (from the v1 api docs anyway), we can only switch zones on/off, not individual devices, which is why I implemented it as a zone..

Can you clarify please?

zxdavb replied:
Specifically, `async_schedule_update_ha_state` is not required.

Also, use `HA_HVAC_TO_GH`: have a look at, say a similar method in **climate.py**::
```python
    async def async_set_hvac_mode(self, hvac_mode: str) -> None:
        """Set a new hvac mode."""
        await self._zone.set_mode(HA_HVAC_TO_GH.get(hvac_mode))
```

castaway replied:
Ah! Ok, what does the update_ha_state bit do ? I assumed it was updating the home assistant ui or similar..)

zxdavb replied:
```python
 from homeassistant.const import (
    STATE_ON,
...

    self._zone.set_mode(HA_HVAC_TO_GH.get(STATE_OFF))
```

zxdavb replied:
> Ah! Ok, what does the update_ha_state bit do ? I assumed it was updating the home assistant ui or similar..)

HA does this automatically, since you did something that made it call the method.

castaway replied:
Hmm, I saw the HVAC -> GH conversions in climate.py, but figured that since I'm not converting from user input (switch ui only has an on/off switch) to internal values, using them would be a bit over the top.. 

Long term it would be handy if the outlets could also set the override time, though I assume this would involve creating something that isn't quite a pure SwitchDevice on the HA end?

zxdavb replied:
> would be a bit over the top

Yes, you're right, sorry, I was being pedantic.

> Long term it would be handy if the outlets could also set the override time

Changing the HA UI would take a lot of energy (you start by raising an issue in the architecture repo), however, it can be done for Automatons, etc., with service calls.

It is in my long-term plans to implement these (and other features), but I just don't have the spare time...

zxdavb commented on homeassistant/components/geniushub/switch.py:
```diff
@@ -0,0 +1,51 @@
+        await self._zone.set_mode("override")
+        # State is set optimistically in the command above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state()
```
My understanding is that this is not required.

zxdavb commented on homeassistant/components/geniushub/switch.py:
```diff
@@ -0,0 +1,51 @@
+    switches = [
+        GeniusSwitch(broker, z)
+        for z in broker.client.zone_objs
+        if GH_ZONE_ATTR in z.data["type"]
```
`z.data["type"] == GH_ON_OFF_ZONE`

zxdavb commented on homeassistant/components/geniushub/switch.py:
```diff
@@ -0,0 +1,51 @@
+        """Return the status of the sensor."""
+        # off, override, timer, footprint
+        # technially this could be untrue, if we're using the timer
+        return self._zone.data["mode"] != "off"
```
It is either (these are both boolean):
```python
self._zone.data["_state"].get("bIsActive"):
```
... or:
```
self._zone.data["_state"].get("bOutRequestHeat"):
```
Sorry, I can't recall exactly, but I think it is the second - please check.

castaway replied:
Ah, _state is the unchanged v3 response? On my outlet zone which is currently on, it responds

`            "bHeatEnabled": true,
            "bIsActive": true,
            "bOutRequestHeat": false,
            "bPrimaryZone": true,
`

zxdavb replied:
Yes, must be `bIsActive`, but actually, I just realized that this limits us to the v3 API - this is inappropriate.

... let me think on it.

castaway replied:
> The problem is, and this is why I haven't been rushing to implement this functionality, that HA switches have no concept of operating mode / profile. That is, they expect to be turned off, and turned on - they have no concept of following a schedule (Timer mode)...

Ahh fun, yes does look like we'd need to expand bits of HA or create a new base thing, to support switches with timers etc. I figured it was nice to have something that at least worked .. perfect being the enemy of the good, and all that.

btw would you take a patch to ghclient repo, for the "set_mode("override", duration) without temp being mentioned ? seems this is probably required for it to stay on..

zxdavb replied:
> btw would you take a patch to ghclient repo, for the "set_mode("override", duration) without temp being mentioned ? seems this is probably required for it to stay on..

All contributions welcome!

zxdavb commented:
Thanks for your PR - we'll definitely get this useful PR merged.

Please add:
```python
    @property
    def device_class(self):
        """Return the class of this device, from component DEVICE_CLASSES."""
        return DEVICE_CLASS_OUTLET
```

zxdavb reviewed (changes requested):
Please add:
```python
    @property
    def device_class(self):
        """Return the class of this device, from component DEVICE_CLASSES."""
        return DEVICE_CLASS_OUTLET
```

zxdavb commented on homeassistant/components/geniushub/switch.py:
```diff
@@ -0,0 +1,51 @@
+    async_add_entities(switches, update_before_add=True)
+
+
+class GeniusSwitch(GeniusZone, SwitchDevice):
```
Unfortunately, you can't just inherit from `GeniusZone` - it is geared up for Climate/WaterHeater entities (e.g. it has `elf._max_temp = self._min_temp = self._supported_features = None`).

You'll have to split `GeniusZone` into two, or extend `GeniusDevice`...

castaway replied:
I see what you mean, on the other hand, they all seem to be zones in geniushub land (at least that's how my install works..) how about we have GeniusZone as a base, and GeniusZoneHeating / GeniusZoneOutlet ?

zxdavb replied:
Hmm... I don't think so - it adds two more classes, and another level of inheritance.

castaway replied:
Ok, you da boss (and python definitely isnt my forte)

zxdavb replied:
Actually, I just changed my mind (I do like a bit of inheritance)!

You're right - spilt GeniusZone in two: GeniusZone  & GeniusHeatingZone.

Then: 
 - GeniusHeatingZone, GeniusSwitch inherit from GeniusZone
 - GeniusClimate, GeniusWaterHeater inherit from GeniusHeatingZone

castaway replied:
Just after I uploaded the other one.. ;)

castaway replied:
.. redone..

zxdavb commented:
> The geniushub integration shows outlets as binary sensors, it can report when they are on or off, but does not allow the user to change the status. This PR adds them as SwitchDevices ...

Please be aware of the distinction between a GH Device (that is a switch), and a GH on/off Zone.

For example, an on/off Zone can have (say) three Devices (smart plugs). All 3 will turn on when the Zone is on, and all three off otherwise.

castaway commented on homeassistant/components/geniushub/switch.py:
```diff
@@ -0,0 +1,51 @@
+        if GH_ZONE_ATTR in z.data["type"]
+    ]
+
+    async_add_entities(switches, update_before_add=True)
```
You're right, I did, and then switched out a bunch of bits from climate.. I've also just noticed that if you start from when the outlet is off, and hit the button - it blinks on, then off again immediately. (ghclient doesn't send a time when purely setting mode, so I may need to patch that too)

If the outlet (ours is a light) is on with override, you can use this to turn it off.. then on again and the existing override continues.

castaway commented:
> > The geniushub integration shows outlets as binary sensors, it can report when they are on or off, but does not allow the user to change the status. This PR adds them as SwitchDevices ...
> 
> Please be aware of the distinction between a GH Device (that is a binary_sensor)), and a GH on/off Zone (a switch).
> 
> For example, an on/off Zone can have (say) three Devices (smart plugs). All 3 will turn on when the Zone is on, and all three off otherwise.

I am aware of these however, (see above), I can only see ways to turn on/off whole zones, not individual devices.. At least in the v1 API.

castaway commented:
> Please add:
> 
> ```python
>     @property
>     def device_class(self):
>         """Return the class of this device, from component DEVICE_CLASSES."""
>         return DEVICE_CLASS_OUTLET
> ```

Aha ok, I took the switch parts from the smartthings/switch.py, which doesn't contain this..

castaway committed:
Tidy up based on some requests from @zxdavb

castaway committed:
Tidy up based on some requests from @zxdavb

zxdavb commented:
> I am aware of these however, (see above), I can only see ways to turn on/off whole zones, not individual devices.. At least in the v1 API.

Correct.

castaway committed:
Fix syntax typo

castaway committed:
Inherit from GeniusEntity, so we don't drag zone heating stuff into Switches

castaway committed:
Inherit from GeniusEntity, so we don't drag zone heating stuff into Switches

castaway committed:
Remove calls, suggestion from @zxdavb

zxdavb commented:
The problem is, and this is why I haven't been rushing to implement this functionality, that HA switches have no concept of operating mode / profile. That is, they expect to be turned off, and turned on - they have no concept of following a schedule (Timer mode)...

castaway commented:
Latest push: https://github.com/home-assistant/home-assistant/pull/28182/commits/857dce085c456c15f275fc7d4edbeb96dfb593ab - should contain all of the suggestions, apart from the discussion on "!= "off" .

castaway committed:
Re-organise to GeniusZone, GeniusHeatingZone, etc

castaway committed:
Re-organise to GeniusZone, GeniusHeatingZone, etc

castaway committed:
Fixed issue with incorrect import
* In which I learn to run pylint locally..

castaway committed:
Fixed issue with incorrect import
* In which I learn to run pylint locally..

zxdavb committed:
de-lint

zxdavb committed:
Fix small bug

zxdavb commented:
FYI: just identified an [issue](https://github.com/zxdavb/geniushub-client/issues/25) with the geniushub-client, where is can't change the override on a on/off zone.

castaway commented:
I felt like crap all weekend... looks like you were busy, what else do we need to poke to make this work as an initial version?

I think the set override time (without temp) on the client is needed, as otherwise the on/off only blinks it on for a second or two.

Anything else?

zxdavb commented:
> Anything else?

Sorry, got side tracked with another PR (now done).

The changes are here: https://github.com/zxdavb/geniushub-client/compare/v0.6.28...toggle_on_off_zones

It previous had a `duration `of 0 seconds for `override` mode - it is now 3600 seconds.

castaway commented:
Ah awesome, much better!

On 29 October 2019 21:49:42 GMT+00:00, David Bonnes <notifications@github.com> wrote:
>> Anything else?
>
>Sorry, got side tracked with another PR (now done).
>
>The changes are here:
>https://github.com/zxdavb/geniushub-client/compare/v0.6.28...toggle_on_off_zones
>
>It previous had a `duration `of 0 seconds for `override` mode - it is
>now 3600 seconds.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/home-assistant/home-assistant/pull/28182#issuecomment-547644937

-- 
Sent from my Android phone with K-9 Mail. Please excuse my brevity.

zxdavb commented:
Please edit **manifest.json** to bump the client (NB: **0.6.29**):
```json
{
  "domain": "geniushub",
  "name": "Genius Hub",
  "documentation": "https://www.home-assistant.io/integrations/geniushub",
  "requirements": [
    "geniushub-client==0.6.29"
  ],
  "dependencies": [],
  "codeowners": ["@zxdavb"]
}
```
... then run: `python3 -m script.gen_requirements_all`

castaway commented:
> Please edit **manifest.json** to bump the client (NB: **0.6.29**):
> 
> ```json
> {
>   "domain": "geniushub",
>   "name": "Genius Hub",
>   "documentation": "https://www.home-assistant.io/integrations/geniushub",
>   "requirements": [
>     "geniushub-client==0.6.29"
>   ],
>   "dependencies": [],
>   "codeowners": ["@zxdavb"]
> }
> ```
> 
> ... then run: `python3 -m script.gen_requirements_all`

I was literally about to look how to do that, well mind read..

castaway commented:
> > Please edit **manifest.json** to bump the client (NB: **0.6.29**):
> > ```json
> > {
> >   "domain": "geniushub",
> >   "name": "Genius Hub",
> >   "documentation": "https://www.home-assistant.io/integrations/geniushub",
> >   "requirements": [
> >     "geniushub-client==0.6.29"
> >   ],
> >   "dependencies": [],
> >   "codeowners": ["@zxdavb"]
> > }
> > ```
> > 
> > 
> > ... then run: `python3 -m script.gen_requirements_all`
> 
> I was literally about to look how to do that, well mind read..

Hrm, except that updated the reqs.. if I tell it then to install from requirements_all.txt (or just run it), it doesnt update.. I did pip3 install geniushub-client==0.6.29 instead

castaway commented:
Hrm. Didnt work - that is, still doesnt set override to an hour.. retested with just latest client from repo to make sure.. doesnt with ghclient either.. time to debug (over on the client repo)

zxdavb commented:
> Hrm, except that updated the reqs.. if I tell it then to install from requirements_all.txt (or just run it), it doesnt update.. I did pip3 install geniushub-client==0.6.29 instead

HA installs the updated client library from PyPI next time you start HA.

Confirm this by removing geniushub-client, and restarting HA:
```
pip uninstall -y geniushub-client
pip list | grep genius
hass

...

pip list | grep genius
```

> Hrm. Didnt work - that is, still doesnt set override to an hour.. retested with just latest client from repo to make sure.. doesnt with ghclient either.. time to debug (over on the client repo)

The script in this issue works for me - have a go at it: https://github.com/zxdavb/geniushub-client/issues/25

You can push your latest PR, and I'll have a look.

castaway commented:
> > Hrm, except that updated the reqs.. if I tell it then to install from requirements_all.txt (or just run it), it doesnt update.. I did pip3 install geniushub-client==0.6.29 instead
> 
> HA installs the updated client library from PyPI next time you start HA.
> 
> Confirm this by removing geniushub-client, and restarting HA:
> 
> ```
> pip uninstall -y geniushub-client
> pip list | grep genius
> hass
> 
> ...
> 
> pip list | grep genius
> ```

Aha, the "remove existing version" wasnt obvious, I assumed it would upgrade if given a higher version.. anyhoo manually got there in the end.

> > Hrm. Didnt work - that is, still doesnt set override to an hour.. retested with just latest client from repo to make sure.. doesnt with ghclient either.. time to debug (over on the client repo)
> 
> The script in this issue works for me - have a go at it: [zxdavb/geniushub-client#25](https://github.com/zxdavb/geniushub-client/issues/25)
> 
> You can push your latest PR, and I'll have a look.

I haven't changed it.. 

Hm, the ghclient.py one? Aha! I'm just calling --zone=X --mode=override .. which still doesn't send a duration (and I think ought to).

castaway commented:
> 
> I haven't changed it..
> 
> Hm, the ghclient.py one? Aha! I'm just calling --zone=X --mode=override .. which still doesn't send a duration (and I think ought to).

To make life more interesting / weird .. now I've tried --zone=9 -t0 -s3600.. and now the light is "Override: Off, override duration: 01:00" in the genius UI

zxdavb commented:
> Aha, the "remove existing version" wasnt obvious, I assumed it would upgrade if given a higher version.. anyhoo manually got there in the end.

You shouldn't need to - I just find that it is often necessary in a dev environment.  Can you confirm which version you're running?

> I haven't changed it..

If you've updated **manifest.json**, then you _have_ changed it.

> To make life more interesting / weird .. now I've tried `--zone=9 -t0 -s3600` .. and now the light is "Override: Off, override duration: 01:00" in the genius UI

To be clear, for an on/off zone, `-t0` is off, `-t1` is on, and the absence of `-s` is the same as `-s3600`.

zxdavb commented:
```
python ghclient.py ${HUB_ADDRESS} -u ${USERNAME} -p ${PASSWORD} -z11 -v
{"id": 11, "name": "WiFi Repeater", "type": "on / off", "mode": "override", "setpoint": false, "override": {"duration": 865, "setpoint": false}}

python ghclient.py ${HUB_ADDRESS} -u ${USERNAME} -p ${PASSWORD} -z11 -t1

python ghclient.py ${HUB_ADDRESS} -u ${USERNAME} -p ${PASSWORD} -z11 -v
{"id": 11, "name": "WiFi Repeater", "type": "on / off", "mode": "override", "setpoint": true, "override": {"duration": 3589, "setpoint": true}}
```

zxdavb committed:
Update switch.py

This is the best we're going to get, given that switches don't have operating modes/schedules.

zxdavb committed:
Update switch.py

Doing it this way may be useful in future, but the HA UI has no way of specifying duration.

zxdavb committed:
de-lint

zxdavb commented:
Please do : https://github.com/home-assistant/home-assistant/pull/28182#issuecomment-548006864

and add the two files to this PR.

castaway commented:
This works! ;)

zxdavb commented:
> This works! ;)

We cannot merge until you update the manifest file as per: https://github.com/home-assistant/home-assistant/pull/28182#issuecomment-548006864

castaway committed:
Update to new gh client with switch on/off support

castaway commented:
done!

castaway committed:
Update requirements for ghclient updated version

zxdavb commented:
I will probably bump the client again before merging.

zxdavb committed:
bump geniushub-client to 0.6.30

zxdavb committed:
bump geniushub-client to 0.6.30

zxdavb commented:
Woops - my fault: sorted now.

zxdavb commented:
Well done! We can merge as soon as the docs PR is sorted.

You're not interested in adding **services** are you?
https://developers.home-assistant.io/docs/en/dev_101_services.html

castaway commented:
https://github.com/home-assistant/home-assistant.io/pull/11095

zxdavb reviewed (approved):
Ready to go
